### PR TITLE
Initial testify work

### PR DESCRIPTION
**Pull request by unclejack**
This PR vendors `stretchr/testify` and makes api_containers integration cli tests use it.

There's a large number of tests. They need to be converted to testify in batches.

Right now, this is WIP. This PR should at least convert all api integration cli tests after agreeing on the parts of `stretchr/testify` we want to use (assert vs if conditions, other refactoring work).

### CODE REVIEW

**Comment by unclejack** [comment]
ping @tianon @LK4D4 @tiborvass @icecrime

**Comment by tiborvass** [comment]
Thanks @unclejack !

Pinging testify lovers @aluzzardi @vieux: what do you think is the right tradeoff between asserts and ifs (e.g. https://github.com/unclejack/docker/commit/9f10d27306b19dcab73c8920746be688dccabc79#diff-ca97ecb464e792b1086fca2ee8af23f8R39)

File: integration-cli/docker_suites_test.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+// need to be added to a specifc suite should be added to this suite.
+type IntegTestSuite struct {
+	suite.Suite
+}
+
+func TestDockerIntegrationTestSuitei(t *testing.T) {
+	suite.Run(t, new(IntegTestSuite))
+}
+
+func (suite *IntegTestSuite) TearDownTest() {
```
Comment (LK4D4) [inline integration-cli/docker_suites_test.go]: I love this, can we also remove all images, except `scratch`, `busybox` and third which I can't remember. And all dangling too.

**Comment by LK4D4** [comment]
I am +1 to use all packages of testify if that will be more convenient then current situation. Also probably multiple `suites` makes sense, for example for `daemon` tests.

File: integration-cli/docker_suites_test.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	"github.com/stretchr/testify/suite"
+)
+
+// IntegTestSuite is the default test suite. All tests which don't
+// need to be added to a specifc suite should be added to this suite.
+type IntegTestSuite struct {
+	suite.Suite
+}
+
+func TestDockerIntegrationTestSuitei(t *testing.T) {
```
Comment (LK4D4) [inline integration-cli/docker_suites_test.go]: Mispell?

File: integration-cli/docker_api_containers_test.go [hunk 3]
```diff
@@ -139,9 +132,7 @@ func TestContainerApiStartVolumeBinds(t *testing.T) {
 	}
 
 	bindPath, err := ioutil.TempDir(os.TempDir(), "test")
-	if err != nil {
-		t.Fatal(err)
-	}
+	assert.Nil(err)
```
Comment (LK4D4) [inline integration-cli/docker_api_containers_test.go]: I believe that `assert` is not equal to fatal. Is this right?

Comment (vieux) [inline integration-cli/docker_api_containers_test.go]: assert.NoError would be better

**Comment by icecrime** [comment]
So I started looking a bit at testify to see what it does, and I feel it is very wrong, and wouldn't use it for Docker as it is. I submitted a stretchr/testify#128 to fix this kind of unexpected behavior:

* `assert.Equal(t, 2.2, 2) == true` but `assert.Equal(t, 2, 2.2) == false`
* `assert.Equal(t, 'x', "x") == true` but `assert.Equal(t, "x", 'x') == false`

So it's mostly NOT LGTM for me unless we get this fixed.

**Comment by vieux** [comment]
@aluzzardi ^^

**Comment by unclejack** [comment]
@icecrime That's OK, we can avoid using assert right now and do that later.

Being able to use the suites to have setup and teardown for the tests is extremely useful.

**Comment by shykes** [comment]
This is labelled as "design review" but I don't seen any changes to UI or public API. What am I missing?

**Comment by jfrazelle** [comment]
Well I wasn't sure if deciding if we _want_ to use testify counts as "design review" or "code review"

**Comment by tiborvass** [comment]
Sorry @unclejack, I'm closing this until we can agree on whether or not to use testify. I'm personally +1 for this, and would love to use both the test suites as well as the assert functions. If there are some weirdnesses that still need to be fixed on it, let's get those PRs merged upstream.

Please continue discussion below.

**Pull request closed without merging**